orderBook.js

```diff
@@ -0,0 +1,49 @@
+function reconcileOrder(existingBook, incomingOrder) {
+  // if the existing book does not contain any data
+  if (existingBook.length === 0) {
+    // add incoming order to empty book
+    existingBook.push(incomingOrder)
+
+    return existingBook
+  }
+  let newArray = []
+
+  // run through the items in the existing book
+  for (let i = 0; i < existingBook.length; i++) {
+    // turn current item in existingBook into a variable so it can be reused
+    let currentItem = existingBook[i]
+
+    // if the order type is the same add it to the existing book
+    // if the order type is different also add it
+    if (incomingOrder.type !== currentItem.type &&
+      (incomingOrder.price === currentItem.price ||
+      (incomingOrder.type === 'sell' && incomingOrder.price < currentItem.price))) {
+      // order type different and price is same
+      // subtract smaller quantity from both orders
+      let smallerQty =
+        incomingOrder.quantity < currentItem.quantity
+          ? incomingOrder.quantity
+          : currentItem.quantity
+
+      currentItem.quantity = currentItem.quantity - smallerQty
+      incomingOrder.quantity = incomingOrder.quantity - smallerQty
+
+      if (currentItem.quantity === 0) {
```

Comment:
we can move "splice and i--" to  outside the if and else here as its required only once

"
existingBook.splice(i, 1)
i--
        
if (currentItem.quantity !== 0) {
    newArray.push(currentItem)
}
"